### DISCUSSION TITLE
How to manually close flyout?

### DISCUSSION THREAD

**Discussion by lukaszossowski** [AvaloniaUI/Avalonia#10593] [Q&A] [answered]
It is supposed to have an IsOpen property, to which I could bind, but the only properties available now are ShowMode, Placement, OverlayInputPassthroughElement and Content.

I have encountered an issue when using an itemsrepeater instead of itemscontrol, after reloading the items the itemscontrol caused the flyout to hide automatically (since the whole control was reloaded I assume), but with itemsrepeater it stays in place (I assume its because of how virtualization works). I would like to find a workaround to close it manually because there is a significant performance gain with the itemsrepeater.

EDIT: Oh I see that IsOpen has a private setter, that's why it doesn't show up.

So, basically what I want to do is close the flyout when any of the buttons inside of it are pressed. Is that possible?

**Comment by lukaszossowski** [comment, marked as answer]
I managed to get it to work by creating my own subclass of Flyout:

```
 public class MyFlyout : Flyout
    {
        protected override void OnOpened()
        {
            if (Content is Control c)
            {
                foreach (var cb in c.GetVisualDescendants().OfType<CheckBox>())
                {
                    cb.Click += (s,e) => Hide();
                }
            }

            base.OnOpened();
        }
    }
```
Since I treat it as a dirty hack already, I searched for the particular control inside my flyout "template" added the Hide method to its Click event.

**Reply by timunie** [reply]
doesn't .Hide() also work without extending Flyout, I mean with any open Flyout?

**Reply by lukaszossowski** [reply]
I'm not sure I understand what you mean. I have a list of items, each of them has a button with flyout, the flyout contains a pin checkbox. When I pin an item the observable collection (that is bound to the Items property) gets recreated with the proper order. When it happens in ItemsControl, the Flyout automatically disappears, but when I change it to ItemsRepeater, the Flyout remains open even though the items get reloaded. 

Here is my XAML:

```
<ItemsRepeater Items="{Binding CurrentRecords}"  >
	<ItemsRepeater.ItemTemplate>
		<DataTemplate>
			<Grid >
				<TextBlock Text="{Binding Name}" />						
				<Button Name="WTF" Classes="Dots" >
					<Button.Flyout>
						<Flyout ShowMode="TransientWithDismissOnPointerMoveAway" Placement="Top"    >
							<CheckBox IsChecked="{Binding Pinned, Mode=TwoWay}" Classes="Pin"  />
						</Flyout>
					</Button.Flyout>
				</Button>
			</Grid>
		</DataTemplate>
	</ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```
In this example when CurrentRecords get updated, the flyout stays open. If it were an ItemsControl, it would disappear.

As an additional note, though I have not tested this thoroughly, the open flyout still works and it seems to be bound to the item that is physically in the place of the old one. So the old one (let's say 4th from the top) gets pushed to the top of the list when pinned, flyout stays open, I can click the CheckBox again and it pins the item that now occupies the "slot" that is 4th from the top and sends it to the top. 

I'm on preview 4 currently.

**Reply by liufeiwu** [reply]
The event binding needs to be removed

```
public class MyFlyout : Flyout
{
    private CheckBox? _cb;
    
    protected override void OnOpened()
    {
        if (Content is Control c)
        {
            foreach (var cb in c.GetVisualDescendants().OfType<CheckBox>())
            {
                _cb= cb;
                _cb.Click += OnClick;
            }
        }

        base.OnOpened();
    }

    protected override void OnClosed()
    {
        if (_cb != null) _cb.Click -= OnClick;
        base.OnClosed();
    }

    private async void OnClick(object? sender, RoutedEventArgs e)
    {
        await Task.Delay(100); // Update ViewModel（wait）
        Hide();
    }
}
```
